File: develop/game.c
#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Constants
#define WIDTH 800
#define HEIGHT 600
#define FPS 60
#define TILE_WIDTH 32
#define TILE_HEIGHT 16
#define GRID_SIZE 10
#define ANIM_FPS 10
#define MAX_PARTICLES 200  // Fixed pool size

// Colors
SDL_Color BLACK = {0, 0, 0, 255};
SDL_Color RED = {255, 0, 0, 255};
SDL_Color GREEN = {0, 255, 0, 255};
SDL_Color BLUE = {0, 0, 255, 255};
SDL_Color YELLOW = {255, 255, 0, 255};
SDL_Color GRAY = {100, 100, 100, 255};

// Structs
typedef struct {
    int x, y;  // Integer positions for speed
    int vx, vy;  // Integer velocities (scaled by 100)
    int lifespan;  // Frames
    int size;
    SDL_Color color;
    int active;  // 1 = active, 0 = inactive
} Particle;

typedef struct {
    int frame;
    Uint32 last_update;
    int frame_delay;
    int num_frames;
} Animation;

typedef struct {
    int x, y;
    int size;
    float speed;
    int health, max_health;
    int dodge_cooldown;
    int dodge_duration;
    int dodging;
    int attack_cooldown;
    char* boons[10];
    int num_boons;
    char state[20];
    int last_x, last_y;
    Animation anim;
} Player;

typedef struct {
    int x, y;
    int size;
    float speed;
    int health;
    Animation anim;
} Enemy;

// Globals
SDL_Window* window = NULL;
SDL_Renderer* renderer = NULL;
Player player;
Enemy* enemies = NULL;
int num_enemies = 0;
Particle particles[MAX_PARTICLES];  // Fixed pool
int num_active_particles = 0;
int anima = 0;
int grid[GRID_SIZE][GRID_SIZE];
int camera_x = 0, camera_y = 0;
Uint32 last_time = 0;
int running = 1;

// Functions
void init_grid() {
    for (int y = 0; y < GRID_SIZE; y++) {
        for (int x = 0; x < GRID_SIZE; x++) {
            grid[y][x] = 1;
        }
    }
}

void init_player() {
    player.x = WIDTH / 2;
    player.y = HEIGHT / 2;
    player.size = 32;
    player.speed = 5.0f;
    player.health = 100;
    player.max_health = 100;
    player.dodge_cooldown = 0;
    player.dodging = 0;
    player.attack_cooldown = 0;
    player.num_boons = 0;
    strcpy(player.state, "idle");
    player.last_x = player.x;
    player.last_y = player.y;
    player.anim.num_frames = 4;
    player.anim.frame = 0;
    player.anim.last_update = 0;
    player.anim.frame_delay = 1000 / ANIM_FPS;
}

void init_animation(Animation* anim) {
    anim->num_frames = 4;
    anim->frame = 0;
    anim->last_update = 0;
    anim->frame_delay = 1000 / ANIM_FPS;
}

void init_enemies(int count) {
    num_enemies = count;
    enemies = malloc(sizeof(Enemy) * num_enemies);
    for (int i = 0; i < num_enemies; i++) {
        enemies[i].x = rand() % (WIDTH - 20);
        enemies[i].y = rand() % (HEIGHT - 20);
        enemies[i].size = 32;
        enemies[i].speed = 2.0f;
        enemies[i].health = 50;
        init_animation(&enemies[i].anim);
    }
}

void free_enemies() {
    if (enemies) free(enemies);
    num_enemies = 0;
}

void init_particles() {
    num_active_particles = 0;
    for (int i = 0; i < MAX_PARTICLES; i++) {
        particles[i].active = 0;  // All inactive
    }
}

void add_particle(int px, int py, int vx, int vy, int lifespan, int size, SDL_Color color) {
    if (num_active_particles >= MAX_PARTICLES) return;  // Cap
    for (int i = 0; i < MAX_PARTICLES; i++) {
        if (!particles[i].active) {
            particles[i].x = px;
            particles[i].y = py;
            particles[i].vx = vx;
            particles[i].vy = vy;
            particles[i].lifespan = lifespan;
            particles[i].size = size;
            particles[i].color = color;
            particles[i].active = 1;
            num_active_particles++;
            break;
        }
    }
}

void generate_room() {
    free_enemies();
    int num = rand() % 4 + 3;
    init_enemies(num);
}

int iso_x(int gx, int gy) {
    return (gx - gy) * TILE_WIDTH / 2 + camera_x;
}

int iso_y(int gx, int gy) {
    return (gx + gy) * TILE_HEIGHT / 2 + camera_y;
}

void draw_rect_filled(int x, int y, int w, int h, SDL_Color color) {
    SDL_Rect rect = {x, y, w, h};
    SDL_SetRenderDrawColor(renderer, color.r, color.g, color.b, color.a);
    SDL_RenderFillRect(renderer, &rect);
}

void draw_circle_filled(int x, int y, int r, SDL_Color color) {
    draw_rect_filled(x - r, y - r, r * 2, r * 2, color);  // Simplified for speed
}

void draw_room() {
    for (int y = 0; y < GRID_SIZE; y++) {
        for (int x = 0; x < GRID_SIZE; x++) {
            if (grid[y][x] == 1) {
                int ix = iso_x(x, y);
                int iy = iso_y(x, y);
                draw_rect_filled(ix, iy, TILE_WIDTH, TILE_HEIGHT, GRAY);
            }
        }
    }
}

void update_camera() {
    int px = player.x, py = player.y;
    camera_x = WIDTH / 2 - (px - py) * TILE_WIDTH / 2;
    camera_y = HEIGHT / 4 - (px + py) * TILE_HEIGHT / 2;
}

void update_animation(Animation* anim, Uint32 current_time) {
    if (current_time - anim->last_update > anim->frame_delay) {
        anim->frame = (anim->frame + 1) % anim->num_frames;
        anim->last_update = current_time;
    }
}

void player_update_state(Uint32 current_time) {
    if (player.x != player.last_x || player.y != player.last_y) {
        if (strcmp(player.state, "walking") != 0) {
            strcpy(player.state, "walking");
        }
    } else {
        if (strcmp(player.state, "idle") != 0) {
            strcpy(player.state, "idle");
        }
    }
    update_animation(&player.anim, current_time);
    player.last_x = player.x;
    player.last_y = player.y;
}

void player_move(const Uint8* keys) {
    int dx = 0, dy = 0;
    if (keys[SDL_SCANCODE_W]) { dx -= player.speed; dy -= player.speed; }
    if (keys[SDL_SCANCODE_S]) { dx += player.speed; dy += player.speed; }
    if (keys[SDL_SCANCODE_A]) { dx -= player.speed; dy += player.speed; }
    if (keys[SDL_SCANCODE_D]) { dx += player.speed; dy -= player.speed; }
    player.x += dx;
    player.y += dy;
    player.x = fmax(0, fmin(player.x, WIDTH - player.size));
    player.y = fmax(0, fmin(player.y, HEIGHT - player.size));
}

void player_dodge() {
    if (player.dodge_cooldown <= 0) {
        player.dodging = player.dodge_duration;
        player.dodge_cooldown = 60;
        player.speed *= 2;
        SDL_Delay(100);
        player.speed /= 2;
    }
}

void player_attack(Enemy* enemies, int num, int mx, int my, Uint32 current_time) {
    if (player.attack_cooldown <= 0) {
        strcpy(player.state, "attacking");
        // Particles: Cone toward mouse (max 8)
        int dx = mx - player.x, dy = my - player.y;
        int dist = (int)sqrt(dx*dx + dy*dy);
        if (dist > 0) {
            int ndx = (dx * 100) / dist, ndy = (dy * 100) / dist;  // Integer math
            int count = (num_active_particles + 8 < MAX_PARTICLES) ? 8 : MAX_PARTICLES - num_active_particles;
            for (int i = 0; i < count; i++) {
                int angle = (rand() % 100 - 50);  // -0.5 to 0.5 radians * 100
                int speed = rand() % 300 + 200;  // 2-5 pixels * 100
                int vx = (cos(angle / 100.0f) * speed * ndx - sin(angle / 100.0f) * speed * ndy) / 100;
                int vy = (sin(angle / 100.0f) * speed * ndx + cos(angle / 100.0f) * speed * ndy) / 100;
                add_particle(player.x, player.y, vx, vy, rand() % 10 + 10, 5, RED);
            }
        }
        // Damage enemies
        for (int i = 0; i < num; i++) {
            int dx = player.x - enemies[i].x, dy = player.y - enemies[i].y;
            int edist = (int)sqrt(dx*dx + dy*dy);
            if (edist < 100) {
                int damage = 20;
                for (int b = 0; b < player.num_boons; b++) {
                    if (strcmp(player.boons[b], "Boon of Rage") == 0) damage += 5;
                }
                enemies[i].health -= damage;
                if (enemies[i].health <= 0) {
                    enemies[i] = enemies[num - 1];
                    num--;
                    i--;
                }
            }
        }
        player.attack_cooldown = 30;
    }
    num_enemies = num;
}

void player_charge(int mx, int my, Uint32 current_time) {
    if (player.attack_cooldown <= 0) {
        int start_x = player.x, start_y = player.y;
        int dx = mx - player.x, dy = my - player.y;
        int dist = (int)sqrt(dx*dx + dy*dy);
        if (dist > 0) {
            dx = (dx * 100) / dist;
            dy = (dy * 100) / dist;
            player.x += dx;
            player.y += dy;
            // Trail particles (max 5)
            int count = (num_active_particles + 5 < MAX_PARTICLES) ? 5 : MAX_PARTICLES - num_active_particles;
            for (int i = 0; i < count; i++) {
                int t = i * 25;  // 0 to 1 * 100
                int px = start_x + (t * dx) / 100;
                int py = start_y + (t * dy) / 100;
                int vx = (rand() % 200 - 100);  // -1 to 1 * 100
                int vy = (rand() % 200 - 100);
                add_particle(px, py, vx, vy, rand() % 5 + 10, 4, YELLOW);
            }
        }
        player.attack_cooldown = 45;
    }
}

void player_update(Uint32 current_time) {
    player_update_state(current_time);
    if (player.dodging > 0) player.dodging--;
    if (player.dodge_cooldown > 0) player.dodge_cooldown--;
    if (player.attack_cooldown > 0) player.attack_cooldown--;
}

void enemy_update(Enemy* e, int px, int py, Uint32 current_time) {
    int dx = px - e->x, dy = py - e->y;
    float dist = sqrt(dx*dx + dy*dy);
    if (dist > 0) {
        dx = (dx / dist) * e->speed;
        dy = (dy / dist) * e->speed;
        e->x += dx;
        e->y += dy;
    }
    update_animation(&e->anim, current_time);
}

void draw_player() {
    int ix = iso_x(player.x, player.y);
    int iy = iso_y(player.x, player.y);
    SDL_Color color = (player.dodging > 0) ? BLUE : RED;
    draw_rect_filled(ix, iy, player.size, player.size, color);
    if (strcmp(player.state, "walking") == 0) {
        color.g = (player.anim.frame % 2) ? 100 : 255;
        draw_rect_filled(ix + 5, iy + 5, 10, 10, color);
    }
}

void draw_enemy(Enemy* e) {
    int ix = iso_x(e->x, e->y);
    int iy = iso_y(e->x, e->y);
    draw_rect_filled(ix, iy, e->size, e->size, GREEN);
    if (e->anim.frame % 2 == 1) {
        draw_rect_filled(ix + 5, iy + 5, 5, 5, BLACK);
    }
}

void update_particles() {
    num_active_particles = 0;
    for (int i = 0; i < MAX_PARTICLES; i++) {
        if (particles[i].active) {
            particles[i].x += particles[i].vx / 100;  // Integer division
            particles[i].y += particles[i].vy / 100;
            particles[i].lifespan--;
            if (particles[i].lifespan <= 0) {
                particles[i].active = 0;
            } else {
                num_active_particles++;
            }
        }
    }
}

void draw_particles() {
    for (int i = 0; i < MAX_PARTICLES; i++) {
        if (particles[i].active) {
            Particle* p = &particles[i];
            int ix = iso_x(p->x, p->y);
            int iy = iso_y(p->x, p->y);
            draw_circle_filled(ix, iy, p->size, p->color);
        }
    }
}

void choose_boon() {
    char* boons[] = {"Boon of Rage", "Boon of the Gladiator"};
    char* chosen = boons[rand() % 2];
    if (player.num_boons < 10) {
        player.boons[player.num_boons] = malloc(strlen(chosen) + 1);
        strcpy(player.boons[player.num_boons], chosen);
        player.num_boons++;
    }
    printf("Applied %s!\n", chosen);
    anima += 50;
}

int main(int argc, char* argv[]) {
    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
        printf("SDL init failed: %s\n", SDL_GetError());
        return 1;
    }

    window = SDL_CreateWindow("Azeroth's Echo - Optimized Particle Prototype", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, WIDTH, HEIGHT, SDL_WINDOW_SHOWN);
    if (!window) {
        printf("Window creation failed: %s\n", SDL_GetError());
        return 1;
    }

    renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC);
    if (!renderer) {
        printf("Renderer creation failed: %s\n", SDL_GetError());
        return 1;
    }

    init_particles();
    init_grid();
    init_player();
    generate_room();
    Uint32 current_time = SDL_GetTicks();

    SDL_Event e;
    while (running) {
        Uint32 new_time = SDL_GetTicks();
        if (new_time - last_time < 1000 / FPS) {
            SDL_Delay(1000 / FPS - (new_time - last_time));
            continue;
        }
        last_time = new_time;
        current_time = new_time;

        while (SDL_PollEvent(&e)) {
            if (e.type == SDL_QUIT) running = 0;
            if (e.type == SDL_MOUSEBUTTONDOWN) {
                int mx, my;
                SDL_GetMouseState(&mx, &my);
                if (e.button.button == SDL_BUTTON_LEFT) {
                    player_attack(enemies, num_enemies, mx, my, current_time);
                } else if (e.button.button == SDL_BUTTON_RIGHT) {
                    player_charge(mx, my, current_time);
                }
            }
        }

        const Uint8* keys = SDL_GetKeyboardState(NULL);
        player_move(keys);
        if (keys[SDL_SCANCODE_SPACE]) player_dodge();

        player_update(current_time);
        for (int i = 0; i < num_enemies; i++) {
            enemy_update(&enemies[i], player.x, player.y, current_time);
            if (player.dodging == 0) {
                int dx = player.x - enemies[i].x, dy = player.y - enemies[i].y;
                float dist = sqrt(dx*dx + dy*dy);
                if (dist < player.size) player.health -= 1;
            }
        }
        update_particles();

        if (num_enemies == 0) {
            choose_boon();
            generate_room();
        }

        if (player.health <= 0) {
            printf("Run ended! Anima collected: %d\n", anima);
            running = 0;
        }

        update_camera();
        SDL_SetRenderDrawColor(renderer, BLACK.r, BLACK.g, BLACK.b, BLACK.a);
        SDL_RenderClear(renderer);
        draw_room();
        draw_particles();
        draw_player();
        for (int i = 0; i < num_enemies; i++) {
            if (enemies[i].y > player.y) draw_enemy(&enemies[i]);
        }
        for (int i = 0; i < num_enemies; i++) {
            if (enemies[i].y <= player.y) draw_enemy(&enemies[i]);
        }
        draw_rect_filled(10, 10, (player.health * 100 / player.max_health), 10, RED);
        SDL_RenderPresent(renderer);
    }

    for (int b = 0; b < player.num_boons; b++) free(player.boons[b]);
    free_enemies();
    SDL_DestroyRenderer(renderer);
    SDL_DestroyWindow(window);
    SDL_Quit();
    return 0;
}

File: develop/game.py
import pygame
import random
import sys
import math

# Initialize Pygame
pygame.init()

# Constants
WIDTH, HEIGHT = 800, 600
FPS = 60
TILE_WIDTH, TILE_HEIGHT = 32, 16
GRID_SIZE = 10
ANIM_FPS = 10

# Colors
BLACK = (0, 0, 0)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)
YELLOW = (255, 255, 0)  # For Charge particles

# Camera
camera_offset = [0, 0]

# Particle Class
class Particle:
    def __init__(self, pos, velocity, lifespan, size=5, color=RED):
        self.pos = pos[:]
        self.velocity = velocity[:]
        self.lifespan = lifespan  # Frames
        self.size = size
        self.color = color
        self.alpha = 255  # For fading (optional)

    def update(self):
        self.pos[0] += self.velocity[0]
        self.pos[1] += self.velocity[1]
        self.lifespan -= 1
        self.size = max(1, self.size - 0.1)  # Shrink over time
        self.alpha = max(0, self.alpha - 255 // self.lifespan)  # Fade (optional)

    def draw(self, screen):
        iso_x = (self.pos[0] - self.pos[1]) * TILE_WIDTH // 2 + camera_offset[0]
        iso_y = (self.pos[0] + self.pos[1]) * TILE_HEIGHT // 2 + camera_offset[1]
        # Use surface for alpha (optional, requires blend mode)
        surface = pygame.Surface((self.size * 2, self.size * 2), pygame.SRCALPHA)
        pygame.draw.circle(surface, (*self.color, self.alpha), (self.size, self.size), self.size)
        screen.blit(surface, (iso_x - self.size, iso_y - self.size))

# Animation Class (unchanged)
class Animation:
    def __init__(self, sprite_sheet, frames, frame_width, frame_height, fps=ANIM_FPS):
        self.sprite_sheet = sprite_sheet
        self.frames = frames
        self.frame_width = frame_width
        self.frame_height = frame_height
        self.fps = fps
        self.current_frame = 0
        self.last_update = 0
        self.frame_delay = 1000 // fps

    def update(self, current_time):
        if current_time - self.last_update > self.frame_delay:
            self.current_frame = (self.current_frame + 1) % len(self.frames)
            self.last_update = current_time

    def get_frame(self):
        col, row = self.frames[self.current_frame]
        x = col * self.frame_width
        y = row * self.frame_height
        return self.sprite_sheet.subsurface((x, y, self.frame_width, self.frame_height))

# Player Class (Updated for Particles)
class Player:
    def __init__(self):
        self.pos = [WIDTH // 2, HEIGHT // 2]
        self.size = 32
        self.speed = 5
        self.health = 100
        self.max_health = 100
        self.dodge_cooldown = 0
        self.dodge_duration = 10
        self.dodging = 0
        self.attack_cooldown = 0
        self.boons = []
        self.state = 'idle'
        self.last_pos = self.pos[:]
        try:
            sheet = pygame.image.load("warrior_sheet.png").convert_alpha()
        except:
            sheet = pygame.Surface((128, 32))
            sheet.fill(RED)
        self.animations = {
            'idle': Animation(sheet, [(0,0), (1,0), (2,0), (3,0)], 32, 32),
            'walking': Animation(sheet, [(0,1), (1,1), (2,1), (3,1)], 32, 32)
        }
        self.current_anim = self.animations['idle']

    def move(self, keys):
        move_x, move_y = 0, 0
        if keys[pygame.K_w]:
            move_x -= self.speed
            move_y -= self.speed
        if keys[pygame.K_s]:
            move_x += self.speed
            move_y += self.speed
        if keys[pygame.K_a]:
            move_x -= self.speed
            move_y += self.speed
        if keys[pygame.K_d]:
            move_x += self.speed
            move_y -= self.speed
        self.pos[0] += move_x
        self.pos[1] += move_y
        self.pos[0] = max(0, min(self.pos[0], WIDTH - self.size))
        self.pos[1] = max(0, min(self.pos[1], HEIGHT - self.size))

    def update_state(self):
        if self.pos != self.last_pos:
            if self.state != 'walking':
                self.state = 'walking'
                self.current_anim = self.animations['walking']
        else:
            if self.state != 'idle':
                self.state = 'idle'
                self.current_anim = self.animations['idle']
        self.last_pos = self.pos[:]

    def dodge(self):
        if self.dodge_cooldown <= 0:
            self.dodging = self.dodge_duration
            self.dodge_cooldown = 60
            self.speed *= 2
            pygame.time.wait(100)
            self.speed /= 2

    def attack(self, enemies, mouse_pos, particles):
        if self.attack_cooldown <= 0:
            self.state = 'attacking'
            # Spawn 5-10 particles in a cone toward mouse
            direction = [mouse_pos[0] - self.pos[0], mouse_pos[1] - self.pos[1]]
            dist = math.sqrt(direction[0]**2 + direction[1]**2)
            if dist > 0:
                direction = [d / dist for d in direction]
                for _ in range(random.randint(5, 10)):
                    angle = random.uniform(-0.5, 0.5)  # Cone spread
                    speed = random.uniform(2, 5)
                    vel = [math.cos(angle) * speed * direction[0] - math.sin(angle) * speed * direction[1],
                           math.sin(angle) * speed * direction[0] + math.cos(angle) * speed * direction[1]]
                    particles.append(Particle(self.pos, vel, random.randint(10, 20), size=5, color=RED))
            for enemy in enemies[:]:
                dist = math.sqrt((self.pos[0] - enemy.pos[0])**2 + (self.pos[1] - enemy.pos[1])**2)
                if dist < 100:
                    damage = 20
                    if "Boon of Rage" in self.boons:
                        damage += 5
                    enemy.health -= damage
                    if enemy.health <= 0:
                        enemies.remove(enemy)
            self.attack_cooldown = 30

    def charge(self, mouse_pos, particles):
        if self.attack_cooldown <= 0:
            start_pos = self.pos[:]
            direction = [mouse_pos[0] - self.pos[0], mouse_pos[1] - self.pos[1]]
            dist = math.sqrt(direction[0]**2 + direction[1]**2)
            if dist > 0:
                direction = [d / dist * 100 for d in direction]
                self.pos[0] += direction[0]
                self.pos[1] += direction[1]
                # Spawn trail particles
                for i in range(5):  # 5 particles along path
                    t = i / 4
                    pos = [start_pos[0] + t * direction[0], start_pos[1] + t * direction[1]]
                    vel = [random.uniform(-1, 1), random.uniform(-1, 1)]  # Small random spread
                    particles.append(Particle(pos, vel, random.randint(10, 15), size=4, color=YELLOW))
            self.attack_cooldown = 45

    def update(self, current_time):
        self.update_state()
        self.current_anim.update(current_time)
        if self.dodging > 0:
            self.dodging -= 1
        if self.dodge_cooldown > 0:
            self.dodge_cooldown -= 1
        if self.attack_cooldown > 0:
            self.attack_cooldown -= 1

    def draw(self, screen):
        iso_x = (self.pos[0] - self.pos[1]) * TILE_WIDTH // 2 + camera_offset[0]
        iso_y = (self.pos[0] + self.pos[1]) * TILE_HEIGHT // 2 + camera_offset[1]
        frame = self.current_anim.get_frame()
        screen.blit(frame, (iso_x, iso_y))

# Enemy Class (unchanged)
class Enemy:
    def __init__(self, pos):
        self.pos = pos
        self.size = 32
        self.speed = 2
        self.health = 50
        try:
            sheet = pygame.image.load("enemy_sheet.png").convert_alpha()
        except:
            sheet = pygame.Surface((64, 32))
            sheet.fill(GREEN)
        self.anim = Animation(sheet, [(0,0), (1,0)], 32, 32)

    def update(self, player_pos, current_time):
        direction = [player_pos[0] - self.pos[0], player_pos[1] - self.pos[1]]
        dist = math.sqrt(direction[0]**2 + direction[1]**2)
        if dist > 0:
            direction = [d / dist * self.speed for d in direction]
            self.pos[0] += direction[0]
            self.pos[1] += direction[1]
        self.anim.update(current_time)

    def draw(self, screen):
        iso_x = (self.pos[0] - self.pos[1]) * TILE_WIDTH // 2 + camera_offset[0]
        iso_y = (self.pos[0] + self.pos[1]) * TILE_HEIGHT // 2 + camera_offset[1]
        frame = self.anim.get_frame()
        screen.blit(frame, (iso_x, iso_y))

# Room Generation (unchanged)
def generate_room():
    enemies = []
    num_enemies = random.randint(3, 6)
    for _ in range(num_enemies):
        pos = [random.randint(0, WIDTH - 20), random.randint(0, HEIGHT - 20)]
        enemies.append(Enemy(pos))
    return enemies

# Draw Room (unchanged)
room_grid = [[1 for _ in range(GRID_SIZE)] for _ in range(GRID_SIZE)]
def draw_room(screen):
    for y in range(GRID_SIZE):
        for x in range(GRID_SIZE):
            if room_grid[y][x] == 1:
                iso_x = (x - y) * TILE_WIDTH // 2 + camera_offset[0]
                iso_y = (x + y) * TILE_HEIGHT // 2 + camera_offset[1]
                pygame.draw.rect(screen, (100, 100, 100), (iso_x, iso_y, TILE_WIDTH, TILE_HEIGHT))

# Camera Update (unchanged)
def update_camera(player):
    iso_x = (player.pos[0] - player.pos[1]) * TILE_WIDTH // 2
    iso_y = (player.pos[0] + player.pos[1]) * TILE_HEIGHT // 2
    camera_offset[0] = WIDTH // 2 - iso_x
    camera_offset[1] = HEIGHT // 4 - iso_y

# Boons (unchanged)
def choose_boon(player):
    boons = ["Boon of Rage", "Boon of the Gladiator"]
    chosen = random.choice(boons)
    player.boons.append(chosen)
    print(f"Applied {chosen}!")

# Main Game
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Azeroth's Echo - Particle Effects Prototype")
clock = pygame.time.Clock()

player = Player()
enemies = generate_room()
particles = []  # New particle list
anima = 0
current_time = 0

running = True
while running:
    current_time = pygame.time.get_ticks()
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        if event.type == pygame.MOUSEBUTTONDOWN:
            if event.button == 1:
                player.attack(enemies, pygame.mouse.get_pos(), particles)
            if event.button == 3:
                player.charge(pygame.mouse.get_pos(), particles)

    keys = pygame.key.get_pressed()
    player.move(keys)
    if keys[pygame.K_SPACE]:
        player.dodge()

    # Update
    player.update(current_time)
    for enemy in enemies:
        enemy.update(player.pos, current_time)
        if not player.dodging and abs(player.pos[0] - enemy.pos[0]) < player.size and abs(player.pos[1] - enemy.pos[1]) < player.size:
            player.health -= 1
    # Update particles
    for particle in particles[:]:
        particle.update()
        if particle.lifespan <= 0:
            particles.remove(particle)

    # Room clear
    if not enemies:
        choose_boon(player)
        anima += 50
        enemies = generate_room()

    # Death
    if player.health <= 0:
        print(f"Run ended! Anima collected: {anima}")
        running = False

    # Draw
    update_camera(player)
    screen.fill(BLACK)
    draw_room(screen)
    # Draw particles before entities for background effect
    for particle in particles:
        particle.draw(screen)
    entities = [(player.pos[1], player)] + [(enemy.pos[1], enemy) for enemy in enemies]
    entities.sort(key=lambda x: x[0])
    for _, entity in entities:
        entity.draw(screen)
    pygame.draw.rect(screen, RED, (10, 10, player.health / player.max_health * 100, 10))
    pygame.display.flip()
    clock.tick(FPS)

pygame.quit()
sys.exit()